app/src/main/kotlin/compiler/analysis/TypeChecker.kt

```diff
@@ -173,9 +173,30 @@ class TypeChecker(private val nameResolution: Map<Ref<AstNode>, Ref<NamedNode>>,
                         is Function -> report(TypeCheckingError.FunctionAsValue(expression, node))
                     }
                 }
-                is Expression.ArrayElement -> TODO()
-                is Expression.ArrayLength -> TODO()
-                is Expression.ArrayAllocation -> TODO()
+
+                is Expression.ArrayElement -> {
+                    val innerExpressionType = checkExpression(expression.expression) ?: return null // if null then something must've gone bad in the recursion
+
+                    if (innerExpressionType is Type.Array) {
+                        checkExpression(expression.index, Type.Number)
+                        return innerExpressionType.elementType
+                    }
+
+                    report(TypeCheckingError.ExtractionFromNonArrayType(expression.expression, innerExpressionType))
+                }
+
+                is Expression.ArrayLength -> {
+                    val innerExpressionType = checkExpression(expression.expression) ?: return null
+
+                    if (innerExpressionType is Type.Array)
+                        return Type.Number
+
+                    report(TypeCheckingError.LengthOfNonArrayType(expression.expression, innerExpressionType))
+                }
+                is Expression.ArrayAllocation -> {
+                    val elementType = checkExpression(expression.initialization) ?: return null
```

Comment:
Please merge the new strategy of array initialization - this bit should get a bit more complex.